toolchain/driver/clang_runtimes.cpp

```diff
@@ -0,0 +1,403 @@
+  } else if (src_file.ends_with(".cpp")) {
+    args.push_back("-std=c++20");
+  }
+
+  // Collect the additional required flags and dynamic flags for this builder.
+  args.append({
+      "-c",
+      builder_->target_flag_,
+      "-o",
+      obj_path.native(),
+      src_path.native(),
+  });
+  if (!builder_->clang_->RunWithNoRuntimes(args)) {
+    return Error(
+        llvm::formatv("Failed to compile runtime source file '{0}'", src_file));
+  }
+
+  auto obj_result = llvm::NewArchiveMember::getFile(obj_path.native(),
+                                                    /*Deterministic=*/true);
+  if (!obj_result) {
+    return Error(llvm::formatv("Unable to read `{0}` object file: {1}",
+                               src_file,
+                               llvm::fmt_consume(obj_result.takeError())));
+  }
+
+  // Unlink the object file once we've read it. However, we log and ignore
+  // any errors here as they aren't fatal.
+  auto unlink_result = builder_->runtimes_builder_->dir().Unlink(obj_path);
+  if (!unlink_result.ok()) {
+    CARBON_VLOG("Unable to unlink object file `{0}`: {1}\n", obj_path,
+                unlink_result.error());
+  }
+
+  return std::move(*obj_result);
+}
+
+ClangResourceDirBuilder::ClangResourceDirBuilder(
+    ClangRunner* clang, llvm::ThreadPoolInterface* threads,
+    llvm::Triple target_triple, Runtimes* runtimes)
+    : ClangRuntimesBuilderBase(clang, threads, std::move(target_triple)),
+      crt_begin_result_(Error("Never built CRT begin file!")),
+      crt_end_result_(Error("Never built CRT end file!")) {
+  // Ensure we're on a platform where we _can_ build a working runtime.
+  if (target_triple_.isOSWindows()) {
+    result_ =
+        Error("TODO: Windows runtimes are untested and not yet supported.");
+    return;
+  }
+
+  auto build_dir_or_error = runtimes->Build(Runtimes::ClangResourceDir);
+  if (!build_dir_or_error.ok()) {
+    result_ = std::move(build_dir_or_error).error();
+    return;
+  }
+  auto build_dir = *(std::move(build_dir_or_error));
+  if (std::holds_alternative<std::filesystem::path>(build_dir)) {
+    // Found cached build.
+    result_ = std::get<std::filesystem::path>(std::move(build_dir));
+    return;
+  }
+
+  runtimes_builder_ = std::get<Runtimes::Builder>(std::move(build_dir));
+  lib_path_ = std::filesystem::path("lib") / target_triple_.str();
+  archive_.emplace(this, lib_path_ / "libclang_rt.builtins.a",
+                   installation().llvm_runtime_srcs(),
+                   CollectBuiltinsSrcFiles(), /*cflags=*/
+                   llvm::SmallVector<llvm::StringRef>{
+                       "-no-canonical-prefixes",
+                       "-O3",
+                       "-fPIC",
+                       "-ffreestanding",
+                       "-fno-builtin",
+                       "-fomit-frame-pointer",
+                       "-fvisibility=hidden",
+                       "-w",
+                   });
+  tasks_.async([this]() mutable { Setup(); });
```

Comment:
```suggestion
  tasks_.async([this]() { Setup(); });
```
Don't think you need the `mutable` here, since you aren't mutating the capture itself (the this pointer), but the thing it points to.